miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/BarcodableSearchRestController.java

```diff
@@ -0,0 +1,72 @@
+package uk.ac.bbsrc.tgac.miso.webapp.controller.rest;
+
+import static uk.ac.bbsrc.tgac.miso.core.util.LimsUtils.isStringEmptyOrNull;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import uk.ac.bbsrc.tgac.miso.dto.BarcodableDto;
+import uk.ac.bbsrc.tgac.miso.dto.Dtos;
+import uk.ac.bbsrc.tgac.miso.service.BarcodableViewService;
+
+@Controller
+@RequestMapping("/rest/barcodable")
+public class BarcodableSearchRestController extends RestController {
+  private static final String baseUrl = "/miso";
+
+  @Autowired
+  private BarcodableViewService barcodableViewService;
+
+  @RequestMapping(value = "/search", method = RequestMethod.GET)
+  @ResponseBody
+  public List<BarcodableDto> search(@RequestParam(name = "q", required = false) String query) throws IOException {
+    if (isStringEmptyOrNull(query)) {
+      return new ArrayList<>();
+    }
+
+    return barcodableViewService.search(query).stream().map(Dtos::asDto).map(this::addUrl).collect(Collectors.toList());
+  }
+
+  private BarcodableDto addUrl(BarcodableDto dto) {
+    long id = dto.getId();
+
+    switch (dto.getEntityType()) {
```

Comment:
I would modify `EntityType` to have a method like `getUrlFragment()` and then this can be put into the stream as:

```
.map(item -> { BarcodableDto dto = Dtos.asDto(item); dto.setUrl(makeUrl(item.getEntityType().getUrlFragment())); return dto; }).collect
```

Reply:
Yes, I'm not particularly fond of `switch` statements either.  However, putting this logic in `EntityType` would be putting methods in a model object, which I've been trying to avoid.  Would that be preferable?

Reply:
Knowledge and mapping of URLs is the responsibility of the front end and Controllers. I don't think it should go into a model object or enum